File: catalog.rb
# catalog.rb — Ruby версия

require 'json'
require 'date'

DATA_FILE = 'books.json'

class Book
  attr_accessor :id, :isbn, :title, :author, :year, :publisher

  def initialize(id, isbn, title, author, year, publisher)
    @id = id
    @isbn = isbn
    @title = title
    @author = author
    @year = year
    @publisher = publisher
  end

  def to_h
    { id: @id, isbn: @isbn, title: @title, author: @author, year: @year, publisher: @publisher }
  end

  def self.from_h(h)
    new(h[:id], h[:isbn], h[:title], h[:author], h[:year], h[:publisher])
  end
end

class BookCatalog
  attr_reader :books

  def initialize
    @books = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      begin
        data = JSON.parse(File.read(DATA_FILE), symbolize_names: true)
        @books = data.map { |b| Book.from_h(b) }
      rescue
        @books = []
      end
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@books.map(&:to_h)))
  end

  def is_valid_isbn?(isbn)
    isbn = isbn.gsub('-', '').gsub(' ', '')
    if isbn.length == 10
      check_isbn10(isbn)
    elsif isbn.length == 13
      check_isbn13(isbn)
    else
      false
    end
  end

  def check_isbn10(isbn)
    return false unless isbn =~ /^\d{9}[\dX]$/
    sum = (0...9).sum { |i| (i + 1) * isbn[i].to_i }
    check = isbn[9]
    sum += check == 'X' ? 100 : 10 * check.to_i
    sum % 11 == 0
  end

  def check_isbn13(isbn)
    return false unless isbn =~ /^\d{13}$/
    sum = isbn.chars.each_with_index.sum do |c, i|
      digit = c.to_i
      i.even? ? digit : 3 * digit
    end
    sum % 10 == 0
  end

  def add_book(isbn, title, author, year, publisher)
    unless is_valid_isbn?(isbn)
      raise "Неверный ISBN"
    end
    if @books.any? { |b| b.isbn == isbn }
      raise "Книга с таким ISBN уже существует"
    end
    id = @books.size + 1
    @books << Book.new(id, isbn, title, author, year, publisher)
    save
    id
  end

  def list_all
    if @books.empty?
      puts "\e[33mКаталог пуст.\e[0m"
      return
    end
    printf "\e[36m%-4s %-15s %-30s %-20s %-6s %-20s\e[0m\n", "ID", "ISBN", "Название", "Автор", "Год", "Издательство"
    puts "-" * 100
    @books.each do |b|
      title = b.title.length > 30 ? b.title[0...30] : b.title
      author = b.author.length > 20 ? b.author[0...20] : b.author
      pub = b.publisher.length > 20 ? b.publisher[0...20] : b.publisher
      printf "%-4d %-15s %-30s %-20s %-6d %-20s\n", b.id, b.isbn, title, author, b.year, pub
    end
  end

  def search_by_isbn(isbn)
    isbn = isbn.gsub('-', '').gsub(' ', '')
    @books.find { |b| b.isbn == isbn }
  end

  def search_by_text(text)
    text = text.downcase
    @books.select { |b| b.title.downcase.include?(text) || b.author.downcase.include?(text) }
  end

  def delete(id)
    found = @books.find { |b| b.id == id }
    if found
      @books.delete(found)
      save
      true
    else
      false
    end
  end

  def edit(id, field, value)
    book = @books.find { |b| b.id == id }
    raise "Книга не найдена" unless book
    case field
    when 'isbn'
      unless is_valid_isbn?(value)
        raise "Неверный ISBN"
      end
      if @books.any? { |b| b.id != id && b.isbn == value }
        raise "ISBN уже используется"
      end
      book.isbn = value
    when 'title'
      book.title = value
    when 'author'
      book.author = value
    when 'year'
      book.year = value.to_i
    when 'publisher'
      book.publisher = value
    else
      raise "Неизвестное поле"
    end
    save
    true
  end

  def stats
    if @books.empty?
      puts "Нет данных."
      return
    end
    authors = Hash.new(0)
    years = Hash.new(0)
    @books.each do |b|
      authors[b.author] += 1
      years[b.year] += 1
    end
    puts "\e[36m📊 Статистика:\e[0m"
    puts "  Всего книг: #{@books.size}"
    puts "  По авторам:"
    authors.sort_by { |_, c| -c }.each do |a, c|
      puts "    #{a}: #{c}"
    end
    puts "  По годам:"
    years.sort.each do |y, c|
      puts "    #{y}: #{c}"
    end
  end
end

def main
  catalog = BookCatalog.new
  loop do
    puts "\n\e[36m📚 Book Catalog (ISBN) (Ruby)\e[0m"
    puts "1. Добавить книгу"
    puts "2. Показать все книги"
    puts "3. Поиск по ISBN"
    puts "4. Поиск по названию/автору"
    puts "5. Удалить книгу"
    puts "6. Редактировать книгу"
    puts "7. Статистика"
    puts "8. Выход"
    print "Выберите действие: "
    choice = gets.chomp
    case choice
    when "1"
      print "ISBN (10 или 13 цифр): "
      isbn = gets.chomp
      print "Название: "
      title = gets.chomp
      print "Автор: "
      author = gets.chomp
      print "Год: "
      year = gets.chomp.to_i
      print "Издательство: "
      publisher = gets.chomp
      begin
        id = catalog.add_book(isbn, title, author, year, publisher)
        puts "\e[32m✅ Книга добавлена (ID: #{id})\e[0m"
      rescue => e
        puts "\e[31m❌ Ошибка: #{e.message}\e[0m"
      end
    when "2"
      catalog.list_all
    when "3"
      print "Введите ISBN: "
      isbn = gets.chomp
      book = catalog.search_by_isbn(isbn)
      if book
        puts "ID: #{book.id}\nISBN: #{book.isbn}\nНазвание: #{book.title}\nАвтор: #{book.author}\nГод: #{book.year}\nИздательство: #{book.publisher}"
      else
        puts "\e[33mКнига не найдена.\e[0m"
      end
    when "4"
      print "Введите название или автора: "
      text = gets.chomp
      results = catalog.search_by_text(text)
      if results.any?
        results.each { |b| puts "#{b.id}: #{b.title} | #{b.author} | #{b.year}" }
      else
        puts "\e[33mНичего не найдено.\e[0m"
      end
    when "5"
      catalog.list_all
      print "Введите ID для удаления: "
      id = gets.chomp.to_i
      if catalog.delete(id)
        puts "\e[32m✅ Книга удалена.\e[0m"
      else
        puts "\e[31m❌ Книга не найдена.\e[0m"
      end
    when "6"
      catalog.list_all
      print "Введите ID для редактирования: "
      id = gets.chomp.to_i
      print "Какое поле редактировать (isbn, title, author, year, publisher): "
      field = gets.chomp.downcase
      print "Новое значение: "
      value = gets.chomp
      begin
        catalog.edit(id, field, value)
        puts "\e[32m✅ Книга обновлена.\e[0m"
      rescue => e
        puts "\e[31m❌ Ошибка: #{e.message}\e[0m"
      end
    when "7"
      catalog.stats
    when "8"
      puts "До свидания!"
      break
    else
      puts "\e[31mНеверный выбор.\e[0m"
    end
  end
end

main if __FILE__ == $0

File: catalog.js
// catalog.js — JavaScript версия

const fs = require('fs');
const readline = require('readline');

const DATA_FILE = 'books.json';

class Book {
    constructor(id, isbn, title, author, year, publisher) {
        this.id = id;
        this.isbn = isbn;
        this.title = title;
        this.author = author;
        this.year = year;
        this.publisher = publisher;
    }
}

class BookCatalog {
    constructor() {
        this.books = [];
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            try {
                const data = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
                this.books = data.map(b => new Book(b.id, b.isbn, b.title, b.author, b.year, b.publisher));
            } catch {
                this.books = [];
            }
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.books, null, 2));
    }

    isValidISBN(isbn) {
        isbn = isbn.replace(/-/g, '').replace(/ /g, '');
        if (isbn.length === 10) return this._checkISBN10(isbn);
        if (isbn.length === 13) return this._checkISBN13(isbn);
        return false;
    }

    _checkISBN10(isbn) {
        if (!/^\d{9}[\dX]$/.test(isbn)) return false;
        let sum = 0;
        for (let i = 0; i < 9; i++) {
            sum += (i + 1) * parseInt(isbn[i]);
        }
        const check = isbn[9];
        if (check === 'X') sum += 100;
        else sum += 10 * parseInt(check);
        return sum % 11 === 0;
    }

    _checkISBN13(isbn) {
        if (!/^\d{13}$/.test(isbn)) return false;
        let sum = 0;
        for (let i = 0; i < 13; i++) {
            const digit = parseInt(isbn[i]);
            sum += (i % 2 === 0) ? digit : 3 * digit;
        }
        return sum % 10 === 0;
    }

    addBook(isbn, title, author, year, publisher) {
        if (!this.isValidISBN(isbn)) throw new Error('Неверный ISBN');
        if (this.books.some(b => b.isbn === isbn)) throw new Error('Книга с таким ISBN уже существует');
        const id = this.books.length + 1;
        const book = new Book(id, isbn, title, author, year, publisher);
        this.books.push(book);
        this.save();
        return id;
    }

    listAll() {
        if (this.books.length === 0) {
            console.log('\x1b[33mКаталог пуст.\x1b[0m');
            return;
        }
        console.log('\x1b[36m' + 'ID'.padEnd(4) + 'ISBN'.padEnd(15) + 'Название'.padEnd(30) + 'Автор'.padEnd(20) + 'Год'.padEnd(6) + 'Издательство'.padEnd(20) + '\x1b[0m');
        console.log('-'.repeat(100));
        for (const b of this.books) {
            const title = b.title.length > 30 ? b.title.slice(0, 30) : b.title;
            const author = b.author.length > 20 ? b.author.slice(0, 20) : b.author;
            const pub = b.publisher.length > 20 ? b.publisher.slice(0, 20) : b.publisher;
            console.log(`${String(b.id).padEnd(4)} ${b.isbn.padEnd(15)} ${title.padEnd(30)} ${author.padEnd(20)} ${String(b.year).padEnd(6)} ${pub.padEnd(20)}`);
        }
    }

    searchByISBN(isbn) {
        isbn = isbn.replace(/-/g, '').replace(/ /g, '');
        return this.books.find(b => b.isbn === isbn) || null;
    }

    searchByText(text) {
        text = text.toLowerCase();
        return this.books.filter(b => b.title.toLowerCase().includes(text) || b.author.toLowerCase().includes(text));
    }

    delete(id) {
        const index = this.books.findIndex(b => b.id === id);
        if (index !== -1) {
            this.books.splice(index, 1);
            this.save();
            return true;
        }
        return false;
    }

    edit(id, field, value) {
        const book = this.books.find(b => b.id === id);
        if (!book) throw new Error('Книга не найдена');
        switch (field) {
            case 'isbn':
                if (!this.isValidISBN(value)) throw new Error('Неверный ISBN');
                if (this.books.some(b => b.id !== id && b.isbn === value)) throw new Error('ISBN уже используется');
                book.isbn = value;
                break;
            case 'title': book.title = value; break;
            case 'author': book.author = value; break;
            case 'year': book.year = parseInt(value); break;
            case 'publisher': book.publisher = value; break;
            default: throw new Error('Неизвестное поле');
        }
        this.save();
        return true;
    }

    stats() {
        if (this.books.length === 0) {
            console.log('Нет данных.');
            return;
        }
        const total = this.books.length;
        const authors = {};
        const years = {};
        for (const b of this.books) {
            authors[b.author] = (authors[b.author] || 0) + 1;
            years[b.year] = (years[b.year] || 0) + 1;
        }
        console.log('\x1b[36m📊 Статистика:\x1b[0m');
        console.log(`  Всего книг: ${total}`);
        console.log('  По авторам:');
        for (const [a, c] of Object.entries(authors).sort((a, b) => b[1] - a[1])) {
            console.log(`    ${a}: ${c}`);
        }
        console.log('  По годам:');
        for (const [y, c] of Object.entries(years).sort()) {
            console.log(`    ${y}: ${c}`);
        }
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const catalog = new BookCatalog();

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

async function main() {
    while (true) {
        console.log('\x1b[36m\n📚 Book Catalog (ISBN) (JavaScript)\x1b[0m');
        console.log('1. Добавить книгу');
        console.log('2. Показать все книги');
        console.log('3. Поиск по ISBN');
        console.log('4. Поиск по названию/автору');
        console.log('5. Удалить книгу');
        console.log('6. Редактировать книгу');
        console.log('7. Статистика');
        console.log('8. Выход');
        const choice = await ask('Выберите действие: ');
        switch (choice.trim()) {
            case '1': {
                const isbn = await ask('ISBN (10 или 13 цифр): ');
                const title = await ask('Название: ');
                const author = await ask('Автор: ');
                const year = parseInt(await ask('Год: '));
                const publisher = await ask('Издательство: ');
                try {
                    const id = catalog.addBook(isbn, title, author, year, publisher);
                    console.log(`\x1b[32m✅ Книга добавлена (ID: ${id})\x1b[0m`);
                } catch (e) {
                    console.log(`\x1b[31m❌ Ошибка: ${e.message}\x1b[0m`);
                }
                break;
            }
            case '2': catalog.listAll(); break;
            case '3': {
                const isbn = await ask('Введите ISBN: ');
                const book = catalog.searchByISBN(isbn);
                if (book) {
                    console.log(`ID: ${book.id}\nISBN: ${book.isbn}\nНазвание: ${book.title}\nАвтор: ${book.author}\nГод: ${book.year}\nИздательство: ${book.publisher}`);
                } else {
                    console.log('\x1b[33mКнига не найдена.\x1b[0m');
                }
                break;
            }
            case '4': {
                const text = await ask('Введите название или автора: ');
                const results = catalog.searchByText(text);
                if (results.length) {
                    for (const b of results) {
                        console.log(`${b.id}: ${b.title} | ${b.author} | ${b.year}`);
                    }
                } else {
                    console.log('\x1b[33mНичего не найдено.\x1b[0m');
                }
                break;
            }
            case '5': {
                catalog.listAll();
                const id = parseInt(await ask('Введите ID для удаления: '));
                if (catalog.delete(id)) {
                    console.log('\x1b[32m✅ Книга удалена.\x1b[0m');
                } else {
                    console.log('\x1b[31m❌ Книга не найдена.\x1b[0m');
                }
                break;
            }
            case '6': {
                catalog.listAll();
                const id = parseInt(await ask('Введите ID для редактирования: '));
                const field = await ask('Какое поле редактировать (isbn, title, author, year, publisher): ');
                const value = await ask('Новое значение: ');
                try {
                    catalog.edit(id, field.trim(), value.trim());
                    console.log('\x1b[32m✅ Книга обновлена.\x1b[0m');
                } catch (e) {
                    console.log(`\x1b[31m❌ Ошибка: ${e.message}\x1b[0m`);
                }
                break;
            }
            case '7': catalog.stats(); break;
            case '8':
                console.log('До свидания!');
                rl.close();
                return;
            default:
                console.log('\x1b[31mНеверный выбор.\x1b[0m');
        }
    }
}

main().catch(console.error);
